p4_symbolic/symbolic/typed.h

```diff
@@ -0,0 +1,134 @@
+// Copyright 2020 Google LLC
```

Comment:
Looking at this, I would lean towards not introducing this (though I'm not 100% sure it's the right call).

Advantages:
- Assignments are safer.

Disadvantages:
- It provides a false sense of security. Code can still change the bit-width of a metadata field.
- We have to use assertions to fail, which doesn't match what we do anywhere else
- It violates the style guide for &&, || etc
- It adds complexity (it needs more code).
- TypedExpr is a bad name, because z3::Expr is already typed. I can't think of a good name.

I would recommend to instead define the following function over z3::Expr which return a StatusOr<z3::Expr>:

Assign
And
Ife
...

One question. How often do we actually assign? If it's 10 times, the case for introducing this is much weaker than, if say, it was 100 times.